### DISCUSSION TITLE
Send the test results in JSON format to Xray

### DISCUSSION THREAD

**Discussion by RomainVivien** [cypress-io/cypress#28987] [Questions and Help]
Hello, I use Cypress for my tests and Cucumber to write them. I would like to automate sending my tests in Gherkin format to Jira. Currently, this part is functional, but I am facing difficulties in automatically sending the generated JSON file at the end of the tests. This file is crucial for determining whether the test has failed or not.

Run curl --location -X POST "https://jira.****.com/rest/raven/2.0/export/execution/json?projectKey=****&testPlanKey=****" --header "Authorization: ****" --header "Content-Type: application/json" --data-binary @reports.json

**Comment by portyu9** [comment]
Hi @RomainVivien 

Yes. This is best done from Cypress's **Node process**, after all specs have completed, rather than from a test.

You can use the `after:run` event to read the generated JSON report and POST it to Xray:

```js
const { defineConfig } = require('cypress')
const fs = require('node:fs/promises')
const path = require('node:path')

module.exports = defineConfig({
  e2e: {
    setupNodeEvents(on, config) {
      on('after:run', async () => {
        const report = await fs.readFile(
          path.join(config.projectRoot, 'reports.json')
        )

        const response = await fetch(process.env.XRAY_URL, {
          method: 'POST',
          headers: {
            Authorization: process.env.XRAY_AUTH,
            'Content-Type': 'application/json',
          },
          body: report,
        })

        if (!response.ok) {
          throw new Error(
            `Xray upload failed: ${response.status} ${response.statusText}`
          )
        }
      })
    },
  },
})
```

Keep the Xray URL/token in environment variables rather than in the test code.

The important requirement is that `reports.json` must already have been finalized when `after:run` executes. If your Cucumber/reporting plugin also owns the `after:run` hook, preserve that plugin's handler or perform the upload in a post-run script after its report generation completes.

Please mark this answer as accepted if it helped, thank you.
